### PR TITLE
feat(topology): integrate topology layout

### PR DESCRIPTION
**Pull request by christianvogt**
Related to:
https://jira.coreos.com/browse/ODC-457
https://jira.coreos.com/browse/ODC-617

* Sets up the topology page to hold the graph
* Renders react components as graph nodes
* Overlays a toolbar for the graph controls
* Uses a simple D3 layout for nodes as a simple POC of separating out the layout component from the rest of the graph
* Handles browser / container resizing
* Supports pan / zoom
* integrates with topology data model
* integrates with workload nodes

image

There are some outstanding issues with performance and updates coming from firehose that I'd like to address in separate PRs. This PR gives a good start to the integration but further work will be needed to polish the experience and would cause this PR to grow even larger than it already is.

In reviewing, the `D3ForceDirectedRenderer` is a temporary layout and probably doesn't need to be reviewed too closely as we plan to replace it with a better layout. This layout was used for simplicity of integration.

### CODE REVIEW

File: frontend/public/extend/devconsole/pages/Topology.tsx [hunk 1]
```diff
@@ -1,14 +1,48 @@
...
-);
+export interface TopologyPageProps {
+  match: RMatch<{
+    ns?: string;
+  }>;
+}
+
+const TopologyPage: React.SFC<TopologyPageProps> = ({ match }) => {
+  const namespace = match.params.ns;
+  const title = namespace ? `Topology of ${namespace}` : 'Topology';
```
Comment (serenamarie125) [inline frontend/public/extend/devconsole/pages/Topology.tsx]: @christianvogt the page title should just be "Topology"

**Review by serenamarie125 — CHANGES_REQUESTED** [review]

File: frontend/public/extend/devconsole/routes.tsx [hunk 2]
```diff
@@ -16,14 +17,32 @@ const routes: RouteProps[] = [
...
+            (await import('./pages/Topology' /* webpackChunkName: "devconsole-topology" */))
+              .default
+          }
+        />
+      ),
+    }
+  )))()),
   {
-    path: '/dev/topology',
+    path: '/dev/topology/ns/:ns',
```
Comment (divyanshiGupta) [inline frontend/public/extend/devconsole/routes.tsx]: Why do we need to add this twice?

Comment (christianvogt) [reply frontend/public/extend/devconsole/routes.tsx]: We don't :) Just left over code.

File: frontend/public/extend/devconsole/pages/Topology.tsx (hunk 1 above)
Comment (christianvogt) [reply frontend/public/extend/devconsole/pages/Topology.tsx]: ok

**Comment by christianvogt** [comment]
@karthikjeeyar in order to integrate better with the empty state, I refactored the `TopologyDataController` to include the firehose as we needed the result of the transformation in order to decide when to show the empty state.

@sahil143 I have integrated the workload node with the topology and made some small changes in order to pull the correct data from the model. I'd like to do some refactoring to support a default node and a workload node. As before your change I had a default node available but it's slightly different than the workload node. I'd like to combine the two and allow the `BaseNode` to be used as the default node. I wanted to do minimal changes here to get the workload node functional in the layout so any further changes I'll want to address in a separate PR.

File: frontend/public/extend/devconsole/components/topology/Node/NodeWrapper.tsx [hunk 3]
```diff
@@ -1,142 +1,61 @@
...
+        innerRadius={innerCircleRadius}
+        icon={data.data.builderImage}
+        label={data.name}
+        selected={selected}
+        onSelect={onSelect}
+      />
+      <WorkloadNode
+        innerRadius={workloadNodeInnerRadius}
+        outerRadius={workloadNodeOuterRadius}
+        data={data.data.donutStatus.pods}
```
Comment (karthikjeeyar) [inline frontend/public/extend/devconsole/components/topology/Node/NodeWrapper.tsx]: data.data.donutStatus.pods is not explicit about what the data is,  specify an alias to the props `data: node`.

File: frontend/public/extend/devconsole/components/topology/Node/NodeWrapper.tsx [hunk 4]
```diff
@@ -1,142 +1,61 @@
...
-  }
-}
-
-const Wrapper = (props: any) => {
-  return <NodeWrapper {...props} />;
+import { NodeProps, WorkloadData } from '../topology-types';
+
+const NodeWrapper: React.FC<NodeProps<WorkloadData>> = ({
+  id,
+  data,
```
Comment (karthikjeeyar) [inline frontend/public/extend/devconsole/components/topology/Node/NodeWrapper.tsx]: add an alias to the data prop

**Review by karthikjeeyar — COMMENTED** [review]
@christianvogt, when the graph gets updated with the new data the drag, and drop feature fails to work. Probably by diffing the old state and new state of the graph data and add/remove node accordingly should fix this issue and also improves the performance as well.

**Comment by christianvogt** [comment]
@karthikjeeyar in order to do a data diff we need a consistent Id generated for nodes. This is something I was going to address in another pr. There is also a Todo in graph about blind casting the graph data which is causing node positions to be lost on update. So there are still outstanding problems with updating the graph with new data.

I'll look at why DND fails after update.

File: frontend/public/extend/devconsole/components/topology/Node/NodeWrapper.tsx (hunk 4 above)
Comment (christianvogt) [reply frontend/public/extend/devconsole/components/topology/Node/NodeWrapper.tsx]: changed to `data: workload`

File: frontend/public/extend/devconsole/components/topology/Node/NodeWrapper.tsx (hunk 3 above)
Comment (christianvogt) [reply frontend/public/extend/devconsole/components/topology/Node/NodeWrapper.tsx]: changed to `data: workload`

**Comment by christianvogt** [comment]
I fixed the DND issue which was due to not updating the data in the dom node for d3 to function.

**Comment by karthikjeeyar** [comment]
LGTM

**Review by karthikjeeyar — APPROVED** [review]

**Pull request merged**